ntoskrnl/io/pnpmgr/pnpmgr.c

```diff
@@ -1760,6 +1765,79 @@ IopGetParentIdPrefix(PDEVICE_NODE DeviceNode,
     return Status;
 }
 
+static
+ULONG
+IopValidateID(IN PWCHAR Id,
+              IN SIZE_T MaxIdLen,
+              IN ULONG MaxSeparators,
+              IN BOOLEAN IsMultiSz)
+{
+   PWCHAR PtrSymbol;
+   PWCHAR StringEnd;
+   WCHAR Symbol;
+   ULONG SeparatorsCount = 0;
+   PWCHAR PtrPrevSymbol = NULL;
+
+   PAGED_CODE();
+
+   StringEnd = Id + MaxIdLen;
+
+   for (PtrSymbol = Id; PtrSymbol < StringEnd; PtrSymbol++)
+   {
+      Symbol = *PtrSymbol;
+
+      if (Symbol == UNICODE_NULL)
+      {
+         if (!IsMultiSz || (PtrPrevSymbol && PtrSymbol == PtrPrevSymbol + 1))
+         {
+            if (MaxSeparators == SeparatorsCount ||
+                MaxSeparators == MAX_SEPARATORS_MULTISZ)
+            {
+               return (ULONG)(PtrSymbol - Id) + 1;
+            }
+
+            DPRINT("IopValidateID: SeparatorsCount - %lu, MaxSeparators - %lu\n",
+                   SeparatorsCount, MaxSeparators);
+
+            // FIXME logging
+            return 0;
+         }
+
+         StringEnd += MaxIdLen;
```

Comment:
It looks like this is executed when you specify Ids in a multi_sz.
If so, this looks wrong to me, because even an ID length must be <= MaxIdLen as I understand the documentation, but it can actually be smaller than this. In this case, the next ID following it is a new one, but by incrementing like that you completely modify the StringEnd pointer that doesn't point to the real end of the string.

Reply:
What do you propose instead?

Reply:
If we intend to test more characters this is not as bad as it seems. It will be bad if we intend to test less than what can actually be.